scvi/external/cell2location/_model.py

```diff
@@ -0,0 +1,572 @@
+        self.n_fact = n_fact
+        self.batch_size = batch_size
+        
+        self.var_names = pd.Series(adata.var_names, index=adata.var_names)
+        if var_names_read is None:
+            self.var_names_read = pd.Series(self.var_names, index=self.var_names)
+        else:
+            self.var_names_read = pd.Series(adata.var[var_names_read], index=self.var_names)
+        self.obs_names = pd.Series(adata.obs_names, index=adata.obs_names)
+        
+        if fact_names is None:
+            self.fact_names = pd.Series(['fact_' + str(i) for i in range(self.n_fact)])
+        else:
+            self.fact_names = pd.Series(fact_names)
+            
+        if sample_id is None:
+            self.sample_id = pd.Series(['sample' for i in range(self.n_obs)],
+                                       index=self.obs_names)
+        else:
+            self.sample_id = pd.Series(adata.obs[sample_id], index=self.obs_names)
+            
+        self.fact_filt = None
+        
+        # generate one-hot encoded matrix telling which obs belong to whic samples
+        self.obs2sample_df = pd.get_dummies(self.sample_id)
+        # convert to np.ndarray and register
+        adata.obsm['obs2sample_obsm'] = self.obs2sample_df.values
+        self.n_exper = self.obs2sample_df.shape[1]
+        register_tensor_from_anndata(adata, "obs2sample", "obsm", "obs2sample_obsm")
+        
+
+    def train_pyro_v2(self, max_epochs: Optional[int] = None,
+                   use_gpu: Optional[bool] = None,
+                   train_size: float = 1,
+                   validation_size: Optional[float] = None,
+                   batch_size: Optional[int] = None, lr: float = 0.001,
+                   total_grad_norm_constraint: float = 200,
+                   **kwargs):
+        
+        if use_gpu is True:
+            self.module.cuda()
+        
+        plan_kwargs = {'loss_fn': pyro.infer.Trace_ELBO(), # for some reason JitTrace_ELBO does not work
+                       'optim': pyro.optim.ClippedAdam({'lr': lr,
+                                                        # limit the gradient step from becoming too large
+                                                        'clip_norm': total_grad_norm_constraint})}
+        
+        if batch_size is None:
+            batch_size = self.batch_size
+        
+        self.train(max_epochs=max_epochs, use_gpu=use_gpu,
+                   train_size=train_size, validation_size=validation_size, batch_size=batch_size,
+                   plan_kwargs=plan_kwargs,
+                   plan_class=PyroTrainingPlan)
+    
+    
+    def train_pyro(
+        self,
+        max_epochs: int = 30000,
+        lr: float = 0.001,
+        total_grad_norm_constraint: float = 200,
+        use_gpu: Optional[bool] = None,
+        train_size: float = 1,
+        validation_size: Optional[float] = None,
+        batch_size = None,
+        plan_kwargs: Optional[dict] = None,
+        **kwargs,
+    ):
```

Comment:
Thanks for explaining! I tried doing something similar to that in `train_pyro_v2` but that did not work because not all prior parameter tensors were registered.